### ISSUE TITLE
Last Dagrun is not working.

### ISSUE THREAD

**Issue by zpatel-dev** [apache/airflow#17814]
**Apache Airflow version**:2.1.3

**OS**: Running Airflow in docker container.
NAME="Ubuntu"
VERSION="20.04.2 LTS (Focal Fossa)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 20.04.2 LTS"
VERSION_ID="20.04"
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
VERSION_CODENAME=focal
UBUNTU_CODENAME=focal

**Apache Airflow Provider versions**: 
apache-airflow-providers-amazon==2.1.0
apache-airflow-providers-celery==2.0.0
apache-airflow-providers-cncf-kubernetes==2.0.1
apache-airflow-providers-docker==2.1.0
apache-airflow-providers-elasticsearch==2.0.2
apache-airflow-providers-ftp==2.0.0
apache-airflow-providers-google==5.0.0
apache-airflow-providers-grpc==2.0.0
apache-airflow-providers-hashicorp==2.0.0
apache-airflow-providers-http==2.0.0
apache-airflow-providers-imap==2.0.0
apache-airflow-providers-microsoft-azure==3.1.0
apache-airflow-providers-mysql==2.1.0
apache-airflow-providers-postgres==2.0.0
apache-airflow-providers-redis==2.0.0
apache-airflow-providers-sendgrid==2.0.0
apache-airflow-providers-sftp==2.1.0
apache-airflow-providers-slack==4.0.0
apache-airflow-providers-sqlite==2.0.0
apache-airflow-providers-ssh==2.1.0

**Deployment**: docker-compose

**What happened**: last_dagtuns are not showing up. Here is the error:
```
Python version: 3.6.14
Airflow version: 2.1.3
Node: 078f1a56398c
-------------------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/airflow/.local/lib/python3.6/site-packages/flask/app.py", line 2447, in wsgi_app
    response = self.full_dispatch_request()
  File "/home/airflow/.local/lib/python3.6/site-packages/flask/app.py", line 1952, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/home/airflow/.local/lib/python3.6/site-packages/flask/app.py", line 1821, in handle_user_exception
    reraise(exc_type, exc_value, tb)
  File "/home/airflow/.local/lib/python3.6/site-packages/flask/_compat.py", line 39, in reraise
    raise value
  File "/home/airflow/.local/lib/python3.6/site-packages/flask/app.py", line 1950, in full_dispatch_request
    rv = self.dispatch_request()
  File "/home/airflow/.local/lib/python3.6/site-packages/flask/app.py", line 1936, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "/home/airflow/.local/lib/python3.6/site-packages/airflow/www/auth.py", line 34, in decorated
    return func(*args, **kwargs)
  File "/home/airflow/.local/lib/python3.6/site-packages/airflow/utils/session.py", line 70, in wrapper
    return func(*args, session=session, **kwargs)
  File "/home/airflow/.local/lib/python3.6/site-packages/airflow/www/views.py", line 858, in last_dagruns
    for r in query
  File "/home/airflow/.local/lib/python3.6/site-packages/airflow/www/views.py", line 858, in <dictcomp>
    for r in query
AttributeError: 'NoneType' object has no attribute 'isoformat'
```

**What you expected to happen**: It should show me last Dagrun date time

**How to reproduce it**: Run latest 2.1.3 airflow on Ubuntu in docker. Turn off examples and create your own DAG.
![image](https://user-images.githubusercontent.com/28397002/130640732-a995b21d-481e-4c2b-a668-7993ce01aa45.png)

**Anything else we need to know**: I haven't seen this last run working ever since I have started using Aiflow

**Are you willing to submit a PR?**:  Sure! but pretty new to Airflow. I believe there should be condition to check for NoneType before accessing its attributes.

**Comment by uranusjr** [comment]
The code is in `airflow/www/views.py`. Search for `def last_dagruns` and the error is toward the bottom of the function.

Change this

```python
resp = {
    r.dag_id.replace('.', '__dot__'): {
        "dag_id": r.dag_id,
        "execution_date": r.execution_date.isoformat(),
        "start_date": r.start_date.isoformat(),
    }
    for r in query
}
```

to something like this:

```python
def _datetime_to_string(value: Optional[DateTime]) -> Optional[str]:
    if value is None:
        return None
    return value.isoformat()

resp = {
    r.dag_id.replace('.', '__dot__'): {
        "dag_id": r.dag_id,
        "execution_date": _datetime_to_string(r.execution_date),
        "start_date": _datetime_to_string(r.start_date),
    }
    for r in query
}
```

I had this fix in another PR (not merged to main yet), but if you could submit a PR just for this change, it can likely be merged pretty quickly and included in 2.1.4.

**Comment by argzeal** [comment]
Thanks @uranusjr for quick fix. Now I understand that I have DAG which schedule is set to `None`, that also breaks last run of other dags as well. As soon as DAG is manually triggered at least once it works. I will create PR as soon as I can to accommodate the suggested change.

**Comment by zealpatel1990** [comment]
Thanks @uranusjr for quick fix. Now I understand that I have DAG which schedule is set to None, that also breaks last run of other dags as well. As soon as DAG is manually triggered at least once it works. I will create PR as soon as I can to accommodate the suggested change.

**Comment by david30907d** [comment]
bumped into the same error, thanks @uranusjr for the quick fix.

**Comment by uranusjr** [comment]
Fix included as a part of #17732.

**Comment by uranusjr** [comment]
Note that #17732 will only be included in Airflow 2.2 (it contains a new feature), so if you will need to submit a separate PR for it to be backported to 2.1.

**Comment by bhavaniravi** [comment]
I could recreate this issue locally and found a few more insights
Ran airflow with 5 DAGs(non-example)

1. when no dags are active, the `last_dagrun` API returns an empty dict. The loading issue is only on the UI
2. when the 1st DAG is activated, the `last_dagrun` sends an error until the task is completed
3. 1st DAG ran successfully, and the last run was populated. UI is no longer loading
4.  There was a case where subdag just hung for a long time after a dagrun was created, the API returned 500 error then
